v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/templates/SpannerChangeStreamsToElasticsearchTest.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.elasticsearch.templates;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.teleport.v2.elasticsearch.templates.SpannerChangeStreamsToElasticsearch.DataChangeRecordToJsonFn;
+import java.util.List;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.ColumnType;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.DataChangeRecord;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.Mod;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.ModType;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.TypeCode;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.ValueCaptureType;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/** Test cases for {@link SpannerChangeStreamsToElasticsearch}. */
+public class SpannerChangeStreamsToElasticsearchTest {
+  private static final List<ColumnType> rowType =
+      ImmutableList.of(
+          new ColumnType("UserId", new TypeCode("STRING"), true, 1),
+          new ColumnType("AccountId", new TypeCode("STRING"), true, 2),
+          new ColumnType("LastUpdate", new TypeCode("TIMESTAMP"), false, 3),
+          new ColumnType("Balance", new TypeCode("INT"), false, 4));
+  private static final List<Mod> insertMods =
+      ImmutableList.of(
+          new Mod(
+              "{\"UserId\": \"UserId1\", \"AccountId\": \"AccountId1\"}",
+              null,
+              "{\"LastUpdate\": \"2022-09-27T11:30:00.123456Z\", \"Balance\": 1000}"),
+          new Mod(
+              "{\"UserId\": \"UserId1\", \"AccountId\": \"AccountId2\"}",
+              null,
+              "{\"LastUpdate\": \"2022-09-27T11:30:00.123456Z\", \"Balance\": 3000}"),
```

Comment:
FYI I believe you can mock calls into Spanner if you want to test the pipeline e2e